quantity/core.py

```diff
@@ -0,0 +1,329 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+from __future__ import annotations
+
+import operator
+from dataclasses import dataclass
+from functools import cached_property
+from typing import TYPE_CHECKING
+
+import array_api_compat
+import astropy.units as u
+import numpy as np
+from astropy.units.quantity_helper.converters import UFUNC_HELPERS, converters_and_unit
+
+if TYPE_CHECKING:
+    from typing import Any
+
+
+PYTHON_NUMBER = float | int | complex
+TUPLE_OR_LIST = tuple | list
+
+def has_array_namespace(arg):
+    try:
+        array_api_compat.array_namespace(arg)
+    except TypeError:
+        return False
+    else:
+        return True
+
+
+def get_value_and_unit(arg, default_unit=None):
+    # HACK: interoperability with astropy Quantity.  Have protocol?
+    if isinstance(arg, QUANTITY_LIKE):
+        return arg.value, arg.unit
+    else:
+        assert not hasattr(arg, "unit")
+        return arg, default_unit
+
+
+@dataclass(frozen=True, eq=False)
+class Quantity:
+    value: Any
+    unit: u.UnitBase
+
+    def _to_own_unit(self, value):
+        value_value, value_unit = get_value_and_unit(value, self._dimensionless_unit)
+        return value_unit.to(self.unit, value_value)
+
+    @cached_property
+    def _dimensionless_unit(self):
+        return self.unit / self.unit
+
+    @property
+    def __array_namespace__(self):
+        # TODO: make our own?
+        return np
+
+    def __getitem__(self, item):
+        return self.__class__(self.value.__getitem__(item), self.unit)
```

Comment:
Not sure about the comparative speed, but using `dataclasses.replace` works better for subclasses, e.g. Angle, which have additional attributes.

```suggestion
        return replace(self, value=self.value.__getitem__(item))
```

Reply:
Done